File: app/src/main/java/com/iomt/android/HTTPRequests.java
package com.iomt.android;

import android.content.Context;
import android.util.Log;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Objects;

public class HTTPRequests {
    private static final String TAG = "HTTPRequests";
    private String JWT;
    private String UserId;
    private GsonBuilder builder = new GsonBuilder();
    private Gson gson = builder.create();
    private Context context;

    public HTTPRequests(Context ctx, String jwt, String userId) {
        context = ctx;
        JWT = jwt;
        UserId = userId;
    }

    public HTTPRequests(Context ctx) {
        context = ctx;
    }

    public void send_dev(DeviceInfo dev){
        String postUrl = context.getString(R.string.base_url) + "/devices/register/?token=" + JWT + "&user_id=" + UserId;
        RequestQueue requestQueue = Volley.newRequestQueue(context);

        JSONObject postData = new JSONObject();
        try {
            postData.put("device_id", dev.getAddress());
            postData.put("device_name", dev.getName());
            postData.put("device_type", dev.getDevice_type());
        } catch (JSONException e) {
            e.printStackTrace();
        }

        JsonObjectRequest jsonObjectRequest = new JsonObjectRequest(Request.Method.POST, postUrl, postData, response -> {}, error -> {
            Log.d(TAG, "HTTP Error: " + error);
            error.printStackTrace();
        });
        Log.d(TAG, "Sent");
        requestQueue.add(jsonObjectRequest);
    }

    public void del_dev(DeviceInfo dev){
        String getUrl = context.getString(R.string.base_url) + "/devices/delete/?token=" + JWT + "&user_id=" + UserId + "&id=" + dev.getAddress();
        RequestQueue requestQueue = Volley.newRequestQueue(context);

        JsonObjectRequest jsonObjectRequest = new JsonObjectRequest(Request.Method.GET, getUrl, null, response -> {}, error -> {
            Log.d(TAG, "HTTP Error: " + error);
            error.printStackTrace();
        });
        requestQueue.add(jsonObjectRequest);
    }

    public void getDevices(Action action){
        String getUrl = context.getString(R.string.base_url) + "/devices/get/?token=" + JWT + "&user_id=" + UserId;
        RequestQueue requestQueue = Volley.newRequestQueue(context);

        JsonObjectRequest jsonObjectRequest = new JsonObjectRequest(Request.Method.GET, getUrl, null, response -> {
            try {
                String[] args = new String[1];
                args[0] = response.getString("devices");
                Log.d(TAG, args[0]);
                action.run(args);
            } catch (JSONException e) {
                Log.d(TAG, "JSON Error: " + Objects.requireNonNull(e.getMessage()));
                e.printStackTrace();
            }

        }, error -> {
            Log.d(TAG, "HTTP Error: " + error);
            error.printStackTrace();
        });
        requestQueue.add(jsonObjectRequest);
    }

    public void getDeviceTypes(Action action){
        String getUrl = context.getString(R.string.base_url) + "/devices/types/?token=" + JWT + "&user_id=" + UserId;
        RequestQueue requestQueue = Volley.newRequestQueue(context);

        JsonObjectRequest jsonObjectRequest = new JsonObjectRequest(Request.Method.GET, getUrl, null, response -> {
            try {
                String[] args = new String[1];
                args[0] = response.getString("devices");
                Log.d(TAG, "DevTypes: " + args[0]);
                action.run(args);
            } catch (JSONException e) {
                Log.d(TAG, "JSON Error: " + Objects.requireNonNull(e.getMessage()));
                e.printStackTrace();
            }

        }, error -> {
            Log.d(TAG, "HTTP Error: " + error);
            error.printStackTrace();
        });
        requestQueue.add(jsonObjectRequest);
    }

    public void send_login(String login, String password, SuccessAction success, ErrorAction error){
        String postUrl = context.getString(R.string.base_url) + "/auth/";
        RequestQueue requestQueue = Volley.newRequestQueue(context);

        JSONObject postData = new JSONObject();
        try {
            postData.put("login", login);
            postData.put("password", password);

        } catch (JSONException e) {
            e.printStackTrace();
        }

        JsonObjectRequest jsonObjectRequest = new JsonObjectRequest(Request.Method.POST, postUrl, postData, response -> {
            try {
                String[] args = new String[2];
                args[0] = response.getString("jwt");
                args[1] = response.getString("user_id");
                success.run(args, response.getBoolean("confirmed"));
            } catch (JSONException e) {
                error.run();
                Log.d(TAG, "JSON Error: " + Objects.requireNonNull(e.getMessage()));
                e.printStackTrace();
            }
        }, err -> {
            error.run();
            Log.d(TAG, "HTTP Error: " + err);
            err.printStackTrace();
        });
        Log.d(TAG, "Sent");
        requestQueue.add(jsonObjectRequest);
    }

    public void send_reg(
            String name,
            String surname,
            String patronymic,
            String birthdate,
            String email,
            String mobile,
            String login,
            String password,
            Action successAction,
            ErrorAction errorAction){
        String postUrl = context.getString(R.string.base_url) + "/users/register/";
        RequestQueue requestQueue = Volley.newRequestQueue(context);

        JSONObject postData = new JSONObject();
        try {
            postData.put("name", name);
            postData.put("surname", surname);
            postData.put("patronymic", patronymic);
            postData.put("birthdate", birthdate);
            postData.put("email", email);
            postData.put("phone_number", mobile);
            postData.put("login", login);
            postData.put("password", password);

        } catch (JSONException e) {
            e.printStackTrace();
        }

        JsonObjectRequest jsonObjectRequest = new JsonObjectRequest(Request.Method.POST, postUrl, postData, response -> {
            String err = null;
            try {
                err = response.getString("error");
            } catch (JSONException e) {
                e.printStackTrace();
            }
            String[] args = new String[1];
            args[0] = err;
            successAction.run(args);
        }, error -> {
            errorAction.run();
            Log.d(TAG, "HTTP Error: " + error);
            error.printStackTrace();
        });
        Log.d(TAG, "Sent");
        requestQueue.add(jsonObjectRequest);
    }

    public void getData(Action action){
        String getUrl = context.getString(R.string.base_url) + "/users/info/?token=" + JWT + "&user_id=" + UserId;
        RequestQueue requestQueue = Volley.newRequestQueue(context);

        JsonObjectRequest jsonObjectRequest = new JsonObjectRequest(Request.Method.GET, getUrl, null, response -> {
            try {
                String[] args = new String[8];
                args[0] = String.valueOf(response.getInt("weight"));
                args[1] = String.valueOf(response.getInt("height"));
                args[2] = (String) response.get("birthdate");
                args[3] = (String) response.get("phone_number");
                args[4] = (String) response.get("email");
                args[5] = (String) response.get("name");
                args[6] = (String) response.get("surname");
                args[7] = (String) response.get("patronymic");
                action.run(args);
            } catch (JSONException e) {
                Log.d(TAG, "JSON Error: " + Objects.requireNonNull(e.getMessage()));
                e.printStackTrace();
            }

        }, error -> {
            Log.d(TAG, "HTTP Error: " + error);
            error.printStackTrace();
        });
        requestQueue.add(jsonObjectRequest);
    }

    public void send_data(String name,
                          String surname,
                          String patr,
                          String birthdate,
                          String email,
                          String phone,
                          int weight,
                          int height){
        String postUrl =  context.getString(R.string.base_url) + "/users/info/?token=" + JWT + "&user_id=" + UserId;
        RequestQueue requestQueue = Volley.newRequestQueue(context);
        JSONObject postData = new JSONObject();
        try {
            postData.put("name", name);
            postData.put("surname", surname);
            postData.put("birthdate", birthdate);
            postData.put("height", height);
            postData.put("weight", weight);
            postData.put("email", email);
            postData.put("phone_number", phone);
            postData.put("patronymic", patr);

        } catch (JSONException e) {
            e.printStackTrace();
        }

        JsonObjectRequest jsonObjectRequest = new JsonObjectRequest(Request.Method.POST, postUrl, postData, response -> {}, error -> {
            Log.d(TAG, "HTTP Error: " + error);
            error.printStackTrace();
        });
        Log.d(TAG, "Sent");
        requestQueue.add(jsonObjectRequest);
    }
}

File: app/src/main/java/com/iomt/android/HTTPRequests.kt
package com.iomt.android

import android.content.Context
import android.util.Log
import kotlinx.coroutines.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.logging.HttpLoggingInterceptor
import org.json.JSONObject
import java.io.IOException

@Serializable
data class AuthInfo(
        val jwt: String,
        // snake case due to the format of data
        val user_id: String,
        val confirmed: Boolean,
        val wasFailed: Boolean = false,
)

class HTTPRequests(
        private val context: Context,
        private var jwt: String? = null,
        private var userId: String? = null,
){
    private val loggingInterceptor = HttpLoggingInterceptor {
        Log.d(TAG, it)
    }.apply {
        setLevel(HttpLoggingInterceptor.Level.BASIC)
    }
    private val client = OkHttpClient.Builder().addInterceptor(loggingInterceptor).build()

    private val coroutineForRequests = CoroutineScope(Job())

    fun sendDevice(dev: DeviceInfo) {
        val postUrl = "${context.getString(R.string.base_url)}/devices/register/?token=$jwt&user_id=$userId"
        val json = JSONObject().apply {
            put("device_id", dev.address)
            put("device_name", dev.name)
            put("device_type", dev.device_type)
        }.toString().toRequestBody(JSON)
        val request = Request.Builder().url(postUrl).post(json).build()
        coroutineForRequests.launch {
            val response = client.newCall(request).execute()
            if (response.isSuccessful) {
                Log.d(TAG, "Request <$postUrl> was successfully sent.")
            } else {
                Log.e(TAG, "Got [${response.code}] from server with <$postUrl>.")
            }
        }
    }

    fun deleteDevice(dev: DeviceInfo) {
        val getUrl = "${context.getString(R.string.base_url)}/devices/delete/?token=$jwt&user_id=$userId&id=${dev.address}"
        val request = Request.Builder().url(getUrl).get().build()
        coroutineForRequests.launch {
            val response = client.newCall(request).execute()
            if (response.isSuccessful) {
                Log.d(TAG, "Request <$getUrl> was successfully sent.")
            } else {
                Log.e(TAG, "Got [${response.code}] from server with <$getUrl>.")
            }
        }
    }

    fun getDevices(action: Action) {
        Log.w(TAG, "JWT: $jwt, userID: $userId")
        val getUrl = "${context.getString(R.string.base_url)}/devices/get/?token=$jwt&user_id=$userId"
        val request = Request.Builder().url(getUrl).get().build()
        client.newCall(request).enqueue(object: Callback {
            override fun onResponse(call: Call, response: Response) {
                if (response.isSuccessful) {
                    Log.d(TAG, "Request <$getUrl> was successfully sent.")
                    action.run(arrayOf(JSONObject(response.body!!.string()).getString("devices")))
                } else {
                    Log.e(TAG, "Got [${response.code}] from server with <$getUrl>.")
                }
            }
            override fun onFailure(call: Call, e: IOException) {
                    Log.e(TAG, "Got [${e.localizedMessage}] from server with <$getUrl>.")
            }
        })
    }

    fun getDeviceTypes(action: Action) {
        val getUrl = "${context.getString(R.string.base_url)}/devices/types/?token=$jwt&user_id=$userId"
        val request = Request.Builder().url(getUrl).get().build()
        coroutineForRequests.launch {
            val response = client.newCall(request).execute()
            if (response.isSuccessful) {
                Log.d(TAG, "Request <$getUrl> was successfully sent.")
                action.run(arrayOf(JSONObject(response.body!!.string()).getString("devices")))
            } else {
                Log.e(TAG, "Got [${response.code}] from server with <$getUrl>.")
            }
        }
    }

    fun sendLogin(login: String, password: String) = runBlocking {
        val url = "${context.getString(R.string.base_url)}/auth/"
        val json = JSONObject().apply {
            put("login", login)
            put("password", password)
        }.toString().toRequestBody(JSON)
        val request = Request.Builder()
                .url(url)
                .post(json)
                .addHeader("Content-Type", "application/json")
                .build()
        val deferred: CompletableDeferred<AuthInfo> = CompletableDeferred()
        client.newCall(request).enqueue(object: Callback {
            override fun onResponse(call: Call, response: Response) {
                val stringResponse = response.body?.string()
                if (response.isSuccessful && stringResponse != null) {
                    deferred.complete(Json.decodeFromString(stringResponse))
                } else {
                    Log.e(TAG, "Got [${response.code}] from server with <$url>.")
                    deferred.complete(AuthInfo("", "", false, wasFailed = true))
                }
            }
            override fun onFailure(call: Call, e: IOException) {
                Log.e(TAG, "Got [${e.localizedMessage}] from server with <$url>.")
                deferred.complete(AuthInfo("", "", false, wasFailed = true))
            }
        })
        deferred.await()
    }

    fun sendReg(
        name: String?,
        surname: String?,
        patronymic: String?,
        birthdate: String?,
        email: String?,
        mobile: String?,
        login: String?,
        password: String?,
        successAction: Action,
        errorAction: ErrorAction
    ) {
        val postUrl = "${context.getString(R.string.base_url)}/users/register/"
        val json = JSONObject().apply {
            put("name", name!!)
            put("surname", surname!!)
            put("patronymic", patronymic!!)
            put("birthdate", birthdate!!)
            put("email", email!!)
            put("phone_number", mobile!!)
            put("login", login!!)
            put("password", password!!)
        }.toString().toRequestBody(JSON)
        val request = Request.Builder().url(postUrl).post(json).build()
        client.newCall(request).enqueue(
            object: Callback {
                override fun onResponse(call: Call, response: Response) {
                    if (response.isSuccessful) {
                        Log.d(TAG, "Request <$postUrl> was successfully sent.")
                        val err = JSONObject(response.body!!.string()).getString("error")
                        successAction.run(arrayOf(err))
                    } else {
                        Log.e(TAG, "Got [${response.code}] from server with <$postUrl>.")
                        errorAction.run()
                    }
                }
                override fun onFailure(call: Call, e: IOException) {
                    Log.e(TAG, "Got [${e.localizedMessage}] from server with <$postUrl>.")
                    errorAction.run()
                }
            }
        )
    }

    fun getData(action: Action) {
        val getUrl = "${context.getString(R.string.base_url)}/users/info/?token=$jwt&user_id=$userId"
        val request = Request.Builder()
                .url(getUrl)
                .get()
                .build()
        client.newCall(request).enqueue(object: Callback {
            override fun onResponse(call: Call, response: Response) {
                if (response.isSuccessful) {
                    Log.d(TAG, "Request <$getUrl> was successfully sent.")
                    val jsonStringResponse = JSONObject(response.body!!.string())
                    action.run(arrayOf(
                        jsonStringResponse.getInt("weight").toString(),
                        jsonStringResponse.getInt("height").toString(),
                        jsonStringResponse["birthdate"] as String,
                        jsonStringResponse["phone_number"] as String,
                        jsonStringResponse["email"] as String,
                        jsonStringResponse["name"] as String,
                        jsonStringResponse["surname"] as String,
                        jsonStringResponse["patronymic"] as String,
                    ))
                } else {
                    Log.e(TAG, "Got [${response.code}] from server with <$getUrl>.")
                }
            }

            override fun onFailure(call: Call, e: IOException) {
                Log.e(TAG, "Got [${e.localizedMessage}] from server with <$getUrl>.")
            }
        })
    }

    fun sendData(
        name: String?,
        surname: String?,
        patronymic: String?,
        birthdate: String?,
        email: String?,
        phone: String?,
        weight: Int,
        height: Int
    ) {
        val postUrl = "${context.getString(R.string.base_url)}/users/info/?token=$jwt&user_id=$userId"
        val json = JSONObject().apply {
            put("name", name!!)
            put("surname", surname!!)
            put("birthdate", birthdate!!)
            put("height", height.toString())
            put("weight", weight.toString())
            put("email", email!!)
            put("phone_number", phone!!)
            put("patronymic", patronymic!!)
        }.toString().toRequestBody(JSON)
        val request = Request.Builder().url(postUrl).post(json).build()
        client.newCall(request).enqueue(object: Callback {
            override fun onResponse(call: Call, response: Response) {
                if (response.isSuccessful) {
                    Log.d(TAG, "Request <$postUrl> was successfully sent.")
                } else {
                    Log.e(TAG, "Got [${response.code}] from server with <$postUrl>.")
                }
            }
            override fun onFailure(call: Call, e: IOException) {
                Log.e(TAG, "Got [${e.localizedMessage}] from server with <$postUrl>.")
            }
        })

    }

    fun getDeviceConfig(configId: Long): String = runBlocking {
        val url = "${context.getString(R.string.base_url)}/devices/types?id=$configId&token=$jwt"
        val request = Request.Builder().url(url).get().build()
        val deferred: CompletableDeferred<String> = CompletableDeferred()
        client.newCall(request).enqueue(object: Callback {
            override fun onResponse(call: Call, response: Response) {
                val config = response.body?.string()
                if (response.isSuccessful && response.body != null) {
                    deferred.complete(config ?: "")
                } else {
                    Log.e(TAG, "Got [${response.code}] from server with <$url>.")
                    deferred.complete("")
                }
            }
            override fun onFailure(call: Call, e: IOException) {
                Log.e(TAG, "Got [${e.localizedMessage}] from server with <$url>.")
                deferred.complete("")
            }
        })
        deferred.await()
    }

    companion object {
        private const val TAG = "HTTPRequests"
        private val JSON = "application/json; charset=utf-8".toMediaType()
    }
}